stetho/src/main/java/com/facebook/stetho/server/http/LeakySocket.java

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (c) 2014-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+package com.facebook.stetho.server.http;
+
+import android.net.LocalSocket;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Utility to allow upgrading a normal HTTP connection to a raw socket (for use with the WebSocket
+ * implementation).  This requires that we read the previously buffered data and stitch together
+ * a magic input stream
+ */
+public class LeakySocket {
+  private final LocalSocket mSocket;
+  private final LeakyBufferedInputStream mLeakyInput;
+
+  LeakySocket(LocalSocket socket, LeakyBufferedInputStream leakyInput) {
+    mSocket = socket;
+    mLeakyInput = leakyInput;
+  }
+
+  public InputStream getInput() throws IOException {
+    return new CompositeInputStream(new InputStream[] {
+        new ByteArrayInputStream(mLeakyInput.clearBuffer()),
+        mSocket.getInputStream()
+    });
```

Comment:
1) We don't explicitly need clearBuffer (it could just return the buffer and not clear it) but I designed it this way to be explicit that you are taking over ownership of the original, unbuffered stream (see [2])
2) Yes, I wanted to make it possible to transition from a buffered stream to an unbuffered one as this is used to transfer protocol to a text-based one (HTTP/1.x) to a binary one (WebSocket).  The WebSocket approach wants to do its own buffering, if it buffers at all.  It should therefore be given the original, unbuffered socket.

I did rework this a bit as I re-used the code for the protocol sniffing socket handler in the latest round of changes.  Specifically, this now encapsulates the `getInput()` functionality in `LeakyBufferedInputStream` with simply: `leakAndClearInput()` which returns a composite of the buffer and the unbuffered stream.